### ISSUE TITLE
extend nxdl_to_hdf5.py to generate h5py and nexusformat scripts

### ISSUE THREAD

**Issue by RussBerg** [nexusformat/exampledata#24]
extend nxdl_to_hdf5.py to generate **h5py** and **nexusformat** scripts that will create hdf5 files of all application definitions

**Comment by RussBerg** [comment]
after executing the generated scripts produces the h5 files, runing **cnxvalidate** on each of those files produces the following output with the bulk of the errors coming from "Required links missing" which I think is not an issue because I believe cnxvalidate will not be validating links in the future (I might be wrong)

```
bergr@VM:44 >./runValidate
h5py_NXarchive.h5 passed validation


h5py_NXarpes.h5 passed validation


h5py_NXcanSAS.h5 passed validation


h5py_NXdirecttof.h5 passed validation


definition=NXfluo.nxdl.xml message="Required link energy missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/energy sev=error dataPath=/entry/data/energy dataFile=h5py_NXfluo.h5
definition=NXfluo.nxdl.xml message="Required link data missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/data sev=error dataPath=/entry/data/data dataFile=h5py_NXfluo.h5
2 errors and 0 warnings found when validating h5py_NXfluo.h5


h5py_NXindirecttof.h5 passed validation


h5py_NXiqproc.h5 passed validation


definition=NXlauetof.nxdl.xml message="Required link data missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/data sev=error dataPath=/entry/name/data dataFile=h5py_NXlauetof.h5
definition=NXlauetof.nxdl.xml message="Required link time_of_flight missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/time_of_flight sev=error dataPath=/entry/name/time_of_flight dataFile=h5py_NXlauetof.h5
2 errors and 0 warnings found when validating h5py_NXlauetof.h5


definition=NXmonopd.nxdl.xml message="Required link polar_angle missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/polar_angle sev=error dataPath=/entry/untitled_data/polar_angle dataFile=h5py_NXmonopd.h5
definition=NXmonopd.nxdl.xml message="Required link data missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/data sev=error dataPath=/entry/untitled_data/data dataFile=h5py_NXmonopd.h5
2 errors and 0 warnings found when validating h5py_NXmonopd.h5


definition=NXmx.nxdl.xml message="Cannot even find the starting point of the depends_on chain, !some char data!" nxdlPath=/NXentry/NXsample/name sev=error dataPath=/untitled_entry/untitled_sample/depends_on dataFile=h5py_NXmx.h5
definition=NXmx.nxdl.xml message="Cannot even find the starting point of the depends_on chain, !some char data!" nxdlPath=/NXentry/NXinstrument/NXdetector sev=error dataPath=/untitled_entry/untitled_instrument/untitled_detector/depends_on dataFile=h5py_NXmx.h5
definition=NXmx.nxdl.xml message="Data type mismatch, expected NX_BOOLEAN, got H5T_STD_I32LE" nxdlPath=/NXentry/NXinstrument/NXdetector/distance_derived sev=error dataPath=/untitled_entry/untitled_instrument/untitled_detector/distance_derived dataFile=h5py_NXmx.h5
definition=NXmx.nxdl.xml message="Data type mismatch, expected NX_BOOLEAN, got H5T_STD_I32LE" nxdlPath=/NXentry/NXinstrument/NXdetector/beam_center_derived sev=error dataPath=/untitled_entry/untitled_instrument/untitled_detector/beam_center_derived dataFile=h5py_NXmx.h5
definition=NXmx.nxdl.xml message="Data type mismatch, expected NX_BOOLEAN, got H5T_STD_I32LE" nxdlPath=/NXentry/NXinstrument/NXdetector/angular_calibration_applied sev=error dataPath=/untitled_entry/untitled_instrument/untitled_detector/angular_calibration_applied dataFile=h5py_NXmx.h5
definition=NXmx.nxdl.xml message="Rank mismatch expected 4, found 3" nxdlPath=/NXentry/NXinstrument/NXdetector/angular_calibration sev=error dataPath=/untitled_entry/untitled_instrument/untitled_detector/angular_calibration dataFile=h5py_NXmx.h5
definition=NXmx.nxdl.xml message="Data type mismatch, expected NX_BOOLEAN, got H5T_STD_I32LE" nxdlPath=/NXentry/NXinstrument/NXdetector/flatfield_applied sev=error dataPath=/untitled_entry/untitled_instrument/untitled_detector/flatfield_applied dataFile=h5py_NXmx.h5
definition=NXmx.nxdl.xml message="Rank mismatch expected 4, found 3" nxdlPath=/NXentry/NXinstrument/NXdetector/flatfield sev=error dataPath=/untitled_entry/untitled_instrument/untitled_detector/flatfield dataFile=h5py_NXmx.h5
definition=NXmx.nxdl.xml message="Rank mismatch expected 4, found 3" nxdlPath=/NXentry/NXinstrument/NXdetector/flatfield_error sev=error dataPath=/untitled_entry/untitled_instrument/untitled_detector/flatfield_error dataFile=h5py_NXmx.h5
definition=NXmx.nxdl.xml message="Rank mismatch expected 4, found 3" nxdlPath=/NXentry/NXinstrument/NXdetector/flatfield_errors sev=error dataPath=/untitled_entry/untitled_instrument/untitled_detector/flatfield_errors dataFile=h5py_NXmx.h5
definition=NXmx.nxdl.xml message="Data type mismatch, expected NX_BOOLEAN, got H5T_STD_I32LE" nxdlPath=/NXentry/NXinstrument/NXdetector/pixel_mask_applied sev=error dataPath=/untitled_entry/untitled_instrument/untitled_detector/pixel_mask_applied dataFile=h5py_NXmx.h5
definition=NXmx.nxdl.xml message="Data type mismatch, expected NX_BOOLEAN, got H5T_STD_I32LE" nxdlPath=/NXentry/NXinstrument/NXdetector/countrate_correction_applied sev=error dataPath=/untitled_entry/untitled_instrument/untitled_detector/countrate_correction_applied dataFile=h5py_NXmx.h5
definition=NXmx.nxdl.xml message="Rank mismatch expected 2, found 0" nxdlPath=/NXentry/NXinstrument/NXbeam/incident_polarisation_stokes sev=error dataPath=/untitled_entry/untitled_instrument/untitled_beam/incident_polarisation_stokes dataFile=h5py_NXmx.h5
13 errors and 0 warnings found when validating h5py_NXmx.h5


definition=NXrefscan.nxdl.xml message="Required link data missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/data sev=error dataPath=/entry/data/data dataFile=h5py_NXrefscan.h5
definition=NXrefscan.nxdl.xml message="Required link rotation_angle missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/rotation_angle sev=error dataPath=/entry/data/rotation_angle dataFile=h5py_NXrefscan.h5
definition=NXrefscan.nxdl.xml message="Required link polar_angle missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/polar_angle sev=error dataPath=/entry/data/polar_angle dataFile=h5py_NXrefscan.h5
3 errors and 0 warnings found when validating h5py_NXrefscan.h5


definition=NXreftof.nxdl.xml message="Required link data missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/data sev=error dataPath=/entry/data/data dataFile=h5py_NXreftof.h5
definition=NXreftof.nxdl.xml message="Required link time_binning missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/time_binning sev=error dataPath=/entry/data/time_binning dataFile=h5py_NXreftof.h5
2 errors and 0 warnings found when validating h5py_NXreftof.h5


definition=NXsas.nxdl.xml message="Required link data missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/data sev=error dataPath=/untitled_entry/data/data dataFile=h5py_NXsas.h5
1 errors and 0 warnings found when validating h5py_NXsas.h5


definition=NXsastof.nxdl.xml message="Required link data missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/data sev=error dataPath=/untitled_entry/data/data dataFile=h5py_NXsastof.h5
definition=NXsastof.nxdl.xml message="Required link time_of_flight missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/time_of_flight sev=error dataPath=/untitled_entry/data/time_of_flight dataFile=h5py_NXsastof.h5
2 errors and 0 warnings found when validating h5py_NXsastof.h5


definition=NXscan.nxdl.xml message="Required link data missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/data sev=error dataPath=/untitled_entry/untitled_data/data dataFile=h5py_NXscan.h5
definition=NXscan.nxdl.xml message="Required link rotation_angle missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/rotation_angle sev=error dataPath=/untitled_entry/untitled_data/rotation_angle dataFile=h5py_NXscan.h5
2 errors and 0 warnings found when validating h5py_NXscan.h5


definition=NXSPE.nxdl.xml message="Failed to load application definition" sev=fatal dataPath=/untitled_entry dataFile=h5py_NXspe.h5
1 errors and 0 warnings found when validating h5py_NXspe.h5


h5py_NXsqom.h5 passed validation


h5py_NXstxm.h5 passed validation


definition=NXtas.nxdl.xml message="Required link ei missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/ei sev=error dataPath=/entry/untitled_data/ei dataFile=h5py_NXtas.h5
definition=NXtas.nxdl.xml message="Required link ef missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/ef sev=error dataPath=/entry/untitled_data/ef dataFile=h5py_NXtas.h5
definition=NXtas.nxdl.xml message="Required link en missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/en sev=error dataPath=/entry/untitled_data/en dataFile=h5py_NXtas.h5
definition=NXtas.nxdl.xml message="Required link qh missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/qh sev=error dataPath=/entry/untitled_data/qh dataFile=h5py_NXtas.h5
definition=NXtas.nxdl.xml message="Required link qk missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/qk sev=error dataPath=/entry/untitled_data/qk dataFile=h5py_NXtas.h5
definition=NXtas.nxdl.xml message="Required link ql missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/ql sev=error dataPath=/entry/untitled_data/ql dataFile=h5py_NXtas.h5
definition=NXtas.nxdl.xml message="Required link data missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/data sev=error dataPath=/entry/untitled_data/data dataFile=h5py_NXtas.h5
7 errors and 0 warnings found when validating h5py_NXtas.h5


definition=NXtofnpd.nxdl.xml message="Required link data missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/data sev=error dataPath=/entry/data/data dataFile=h5py_NXtofnpd.h5
definition=NXtofnpd.nxdl.xml message="Required link detector_number missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/detector_number sev=error dataPath=/entry/data/detector_number dataFile=h5py_NXtofnpd.h5
definition=NXtofnpd.nxdl.xml message="Required link time_of_flight missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/time_of_flight sev=error dataPath=/entry/data/time_of_flight dataFile=h5py_NXtofnpd.h5
3 errors and 0 warnings found when validating h5py_NXtofnpd.h5


definition=NXtofraw.nxdl.xml message="Required link data missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/data sev=error dataPath=/entry/data/data dataFile=h5py_NXtofraw.h5
definition=NXtofraw.nxdl.xml message="Required link detector_number missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/detector_number sev=error dataPath=/entry/data/detector_number dataFile=h5py_NXtofraw.h5
definition=NXtofraw.nxdl.xml message="Required link time_of_flight missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/time_of_flight sev=error dataPath=/entry/data/time_of_flight dataFile=h5py_NXtofraw.h5
3 errors and 0 warnings found when validating h5py_NXtofraw.h5


definition=NXtofsingle.nxdl.xml message="Required link data missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/data sev=error dataPath=/entry/data/data dataFile=h5py_NXtofsingle.h5
definition=NXtofsingle.nxdl.xml message="Required link detector_number missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/detector_number sev=error dataPath=/entry/data/detector_number dataFile=h5py_NXtofsingle.h5
definition=NXtofsingle.nxdl.xml message="Required link time_of_flight missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/time_of_flight sev=error dataPath=/entry/data/time_of_flight dataFile=h5py_NXtofsingle.h5
3 errors and 0 warnings found when validating h5py_NXtofsingle.h5


definition=NXtomo.nxdl.xml message="Required link data missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/data sev=error dataPath=/entry/data/data dataFile=h5py_NXtomo.h5
definition=NXtomo.nxdl.xml message="Required link rotation_angle missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/rotation_angle sev=error dataPath=/entry/data/rotation_angle dataFile=h5py_NXtomo.h5
definition=NXtomo.nxdl.xml message="Required link image_key missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/image_key sev=error dataPath=/entry/data/image_key dataFile=h5py_NXtomo.h5
3 errors and 0 warnings found when validating h5py_NXtomo.h5


definition=NXtomophase.nxdl.xml message="Required link data missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/data sev=error dataPath=/entry/data/data dataFile=h5py_NXtomophase.h5
definition=NXtomophase.nxdl.xml message="Required link rotation_angle missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/rotation_angle sev=error dataPath=/entry/data/rotation_angle dataFile=h5py_NXtomophase.h5
2 errors and 0 warnings found when validating h5py_NXtomophase.h5


h5py_NXtomoproc.h5 passed validation


definition=NXxas.nxdl.xml message="Required link energy missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/energy sev=error dataPath=/untitled_entry/untitled_data/energy dataFile=h5py_NXxas.h5
definition=NXxas.nxdl.xml message="Required link absorbed_beam missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/absorbed_beam sev=error dataPath=/untitled_entry/untitled_data/absorbed_beam dataFile=h5py_NXxas.h5
2 errors and 0 warnings found when validating h5py_NXxas.h5


definition=NXxasproc.nxdl.xml message="Rank mismatch expected 1, found 0" nxdlPath=/NXentry/NXdata/energy sev=error dataPath=/untitled_entry/untitled_data/energy dataFile=h5py_NXxasproc.h5
1 errors and 0 warnings found when validating h5py_NXxasproc.h5


definition=NXxbase.nxdl.xml message="Required link data missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/data sev=error dataPath=/entry/untitled_data/data dataFile=h5py_NXxbase.h5
1 errors and 0 warnings found when validating h5py_NXxbase.h5


definition=NXxeuler.nxdl.xml message="Required link polar_angle missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/polar_angle sev=error dataPath=/entry/name/polar_angle dataFile=h5py_NXxeuler.h5
definition=NXxeuler.nxdl.xml message="Required link rotation_angle missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/rotation_angle sev=error dataPath=/entry/name/rotation_angle dataFile=h5py_NXxeuler.h5
definition=NXxeuler.nxdl.xml message="Required link chi missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/chi sev=error dataPath=/entry/name/chi dataFile=h5py_NXxeuler.h5
definition=NXxeuler.nxdl.xml message="Required link phi missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/phi sev=error dataPath=/entry/name/phi dataFile=h5py_NXxeuler.h5
definition=NXxeuler.nxdl.xml message="Required link data missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/data sev=error dataPath=/entry/name/data dataFile=h5py_NXxeuler.h5
5 errors and 0 warnings found when validating h5py_NXxeuler.h5


definition=NXxkappa.nxdl.xml message="Required link polar_angle missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/polar_angle sev=error dataPath=/entry/name/polar_angle dataFile=h5py_NXxkappa.h5
definition=NXxkappa.nxdl.xml message="Required link rotation_angle missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/rotation_angle sev=error dataPath=/entry/name/rotation_angle dataFile=h5py_NXxkappa.h5
definition=NXxkappa.nxdl.xml message="Required link kappa missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/kappa sev=error dataPath=/entry/name/kappa dataFile=h5py_NXxkappa.h5
definition=NXxkappa.nxdl.xml message="Required link phi missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/phi sev=error dataPath=/entry/name/phi dataFile=h5py_NXxkappa.h5
definition=NXxkappa.nxdl.xml message="Required link data missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/data sev=error dataPath=/entry/name/data dataFile=h5py_NXxkappa.h5
5 errors and 0 warnings found when validating h5py_NXxkappa.h5


h5py_NXxlaue.h5 passed validation


h5py_NXxlaueplate.h5 passed validation


definition=NXxnb.nxdl.xml message="Required link polar_angle missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/polar_angle sev=error dataPath=/entry/name/polar_angle dataFile=h5py_NXxnb.h5
definition=NXxnb.nxdl.xml message="Required link tilt missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/tilt sev=error dataPath=/entry/name/tilt dataFile=h5py_NXxnb.h5
definition=NXxnb.nxdl.xml message="Required link rotation_angle missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/rotation_angle sev=error dataPath=/entry/name/rotation_angle dataFile=h5py_NXxnb.h5
definition=NXxnb.nxdl.xml message="Required link data missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/data sev=error dataPath=/entry/name/data dataFile=h5py_NXxnb.h5
4 errors and 0 warnings found when validating h5py_NXxnb.h5


definition=NXxrot.nxdl.xml message="Required link rotation_angle missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/rotation_angle sev=error dataPath=/entry/name/rotation_angle dataFile=h5py_NXxrot.h5
definition=NXxrot.nxdl.xml message="Required link data missing or not pointing to an HDF5 object" nxdlPath=/NXentry/NXdata/data sev=error dataPath=/entry/name/data dataFile=h5py_NXxrot.h5
2 errors and 0 warnings found when validating h5py_NXxrot.h5

**Comment by mkoennecke** [comment]
Hi Russ,

cnxvalidate does validate links. The only thing which was removed was checking the target attribute. Thus a link can 
point anywhere in the file. However, cnxvalidate still checks that the link exists and points to a valid HDF5 object. And those links are an important part of NeXus, especially in NXdata. Thus: can I ask you to fix the link problem?

**Comment by RussBerg** [comment]
Hey @mkoennecke I just got off that other call and can finally run my generated files against the most up to date version of cnxvalidate, I am getting a bunch of new messages, in particular every file now is being checked for a **HDF5_Version** attribute but I can't see where this is declared to be global requirement in the documentation, ex:
```
message="Missing required global HDF5_Version attribute" sev=error dataPath=/ dataFile=NXmonopd.hdf5

**Comment by mkoennecke** [comment]
Russ, 

there has been some discussion on this, see https://github.com/nexusformat/cnxvalidate/issues/24 
Why not create the HDF5_Version attribute? It is highly recommended to have that thing because there are incompatibilities in HDF5: for example HDF 1.8 may not be able to read HDF5 1.10 written file if it uses features not available in 1.8
